apps/src/dance/ai/utils/validateAndSetFieldValue.ts

```diff
@@ -0,0 +1,24 @@
+import {FieldDropdown} from 'blockly';
+import {FieldKey} from '@cdo/apps/dance/ai/types';
+
+import Lab2MetricsReporter from '@cdo/apps/lab2/Lab2MetricsReporter';
+
+export function validateAndSetFieldValue(
+  dropdown: FieldDropdown,
+  value: string,
+  field: FieldKey
+) {
+  if (
+    dropdown
+      .getOptions()
+      .some(option => option[1] === value || option[1] === `"${value}"`)
+  ) {
+    dropdown.setValue(value);
+  } else {
+    Lab2MetricsReporter.logWarning({
```

Comment:
Potential follow-up: I can see this util function being useful outside of this context as well, but in order to do so, we'd probably want to extract out the Lab2-specific logging code - probably instead passing an argument like `onInvalidValue: (value: string) => void`?

Reply:
Veering ever farther of of my original goal for this PR to just shuffle around code 😉 , but nonetheless...

I retooled it. Now it accepts an optional fourth argument - `options`, which contains a key `invalidValueLogger` that contains a function with the appropriate signature that's called when something invalid comes in.

It's in an `options` object to emphasize a distinction between it being an argument required for the _operation_ of the function, and something used to _augment the behavior_ of the function.

Plus, you can get a partially applied version of the function via `const validateAndSetFieldValue = getValidateAndSetFieldValueWithInvalidValueLogger(logger)` so you don't need to keep handing in the logger value. The partially applied version still allows you to pass in a new `invalidValueLogger` in the options, if you want to override it on a case by case basis.

e.g., `generateAiEffectBlocksFromResult.ts` now has this:

```
const validateAndSetFieldValue =
  getValidateAndSetFieldValueWithInvalidValueLogger(
    Lab2MetricsReporter.logWarning.bind(Lab2MetricsReporter)
  );
```

and then proceeds as normal.

Reply:
Interesting! That seems handy - one thing I'm unsure about though, is if the caller wants to supply additional arguments to the logging function, is it still useful to have the `getValidateAndSetFieldValueWithInvalidValueLogger` util? I ask because in order to fully decouple this util, it would be ideal to also remove the reference to `FieldKey` as that's a Dance AI-specific type, while this util could feasibly be used for any Blockly dropdown. The `FieldKey` was only supplied to the function for the logging purposes anyway. In doing so though, we'd still want a way for the other Dance AI util functions to pass the `FieldKey` to the logging function. So I'd imagine we'd end up with something like (in `generateBlocksFromResult`)

```
  // Foreground block.
  validateAndSetFieldValue(
    blocksSvg[0].getField('EFFECT') as FieldDropdown,
    effect.foregroundEffect,
    { 
      invalidValueLogger: (message, value) => {
        Lab2MetricsReporter.logWarning({message, value, field: FieldKey.FOREGROUND_EFFECT});
    }
  );
```

Reply:
Oh! I see. I didn't pay attention to the fact that `field` was only used for logging. 😊 

I re-tooled. The `options` hash now takes two values - `invalidValueLogger` and `logValues` (which is an unknown).

So you can still get the logger as before, though I changed it to remap the generic `logValues` key back to `field`.
```
const validateAndSetFieldValue =
  getValidateAndSetFieldValueWithInvalidValueLogger(
    ({message, value, logValues}) =>
      Lab2MetricsReporter.logWarning({message, value, field: logValues})
  );
```

And the call just changes from this:
```
  validateAndSetFieldValue(
    blocksSvg[0].getField('EFFECT') as FieldDropdown,
    effect.foregroundEffect,
    FieldKey.FOREGROUND_EFFECT
  );
```

to this:
```
  validateAndSetFieldValue(
    blocksSvg[0].getField('EFFECT') as FieldDropdown,
    effect.foregroundEffect,
    {logValues: FieldKey.FOREGROUND_EFFECT}
  );
```

And the partially applied function still works, and we don't need to duplicate the logger all over the place. 😄 

`logValues` is explicitly typed as `unknown`, since it's intended to be an opaque input that's merely passed along to the logger.

Reply:
Cool yep, that works!